docs/architecture/adr-018-dynamic-capability-store.md

```diff
@@ -0,0 +1,197 @@
+  // set reverse mapping in memory store from name to capability
+  memoryStore.Set("rev/" + name, capability)
+  // return the newly created capability
+  return capability
+}
+```
+
+`AuthenticateCapability` can be called by any module to check that a capability
+does in fact correspond to a particular name (the name can be untrusted user input).
+
+```golang
+func (ck CapabilityKeeper) AuthenticateCapability(name string, capability Capability) bool {
+  memoryStore := ctx.KVStore(ck.memoryKey)
+  // return whether forward mapping in memory store matches name
+  return memoryStore.Get("fwd/" + capability) === name
+}
+```
+
+`Initialise` MUST be called exactly once, after loading the initial state, to populate the memory store with newly-created capability keys
+in accordance with the keys previously claimed by particular modules.
+
+```golang
+func (ck CapabilityKeeper) Initialise(ctx sdk.Context) {
+  persistentStore := ctx.KVStore(ck.persistentKey)
+  memoryStore := ctx.KVStore(ck.memoryKey)
+  // initialise memory store for all names in persistent store
+  for key, value := range persistentStore.Iter() {
+    capability = &CapabilityKey{name: key}
+    memoryStore.Set("fwd/" + capability, name)
+    memoryStore.Set("rev/" + name, capability)
+  }
+}
+```
+
+The `CapabilityKeeper` also provides the ability to create *scoped* sub-keepers which are tied to a particular module name. These `ScopedCapabilityKeeper`s must be created at application
+initialisation and passed to modules, which can then use them to claim capabilities they receive and retrieve capabilities which they own by name.
+
+```golang
+type ScopedCapabilityKeeper struct {
+  capabilityKeeper CapabilityKeeper
+  moduleName string
+}
+```
+
+`ScopeToModule` is used to create a scoped sub-keeper with a particular name, which must be unique.
+
+```golang
+func (ck CapabilityKeeper) ScopeToModule (moduleName string) {
+  if _, present := ck.moduleNames[moduleName]; present {
+    panic("cannot create multiple scoped capability keepers for the same module name")
+  }
+  ck.moduleNames[moduleName] = interface{}
+  return ScopedCapabilityKeeper{
+    capabilityKeeper: ck,
+    moduleName: moduleName
+  }
+}
+```
+
+`ClaimCapability` allows a module to claim a capability key which it has received (perhaps by calling `NewCapability`, or from another module), so that future `GetCapability` calls will succeed.
+
+`ClaimCapability` MUST be called, even if `NewCapability` was called by the same module. Capabilities are single-owner, so if multiple modules have a single `Capability` reference, the last module
+to call `ClaimCapability` will own it. To avoid confusion, a module which calls `NewCapability` SHOULD either call `ClaimCapability` or pass the capability to another module which will then claim it.
+
+```golang
+func (sck ScopedCapabilityKeeper) ClaimCapability(ctx sdk.Context, capability Capability) error {
+  persistentStore := ctx.KVStore(sck.ck.persistentKey)
```

Comment:
note, `persistentKey` and `memoryKey` are private so you won't be able to use them as it's presented here.

Reply:
Aye, fixed.